fix capacity issue for category aware

AlbericByte opened the pull request:
<!-- Thanks for trying to help us make Apache Druid be the best it can be! Please fill out as much of the following information as is possible (where relevant, and remove it when irrelevant) to help make the intention and scope of this PR clear in order to ease review. -->

<!-- Please read the doc for contribution (https://github.com/apache/druid/blob/master/CONTRIBUTING.md) before making this PR. Also, once you open a PR, please _avoid using force pushes and rebasing_ since these make it difficult for reviewers to see what you've changed in response to their reviews. See [the 'If your pull request shows conflicts with master' section](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#if-your-pull-request-shows-conflicts-with-master) for more details. -->

Fixes #15847.

<!-- Replace XXXX with the id of the issue fixed in this PR. Remove this section if there is no corresponding issue. Don't reference the issue in the title of this pull-request. -->

<!-- If you are a committer, follow the PR action item checklist for committers:
https://github.com/apache/druid/blob/master/dev/committer-instructions.md#pr-and-issue-action-item-checklist-for-committers. -->

### Description

Fix the bug of capacity api which is not category aware
The proposal solution is following:

1. If current WorkerSelectStrategy is category based strategy(FillCapacityWithCategorySpecWorkerSelectStrategy and EqualDistributionWithCategorySpecWorkerSelectStrategy),  we will calculate the capacity group by category in getWorkerCategoryCapacity method.
2. if current WorkerSelectStrategy is not category based strategy, the getWorkerCategoryCapacity will return null.
3. In OverlordResource#getTotalWorkerCapacity method, get the capacity group by category from WorkerSelectStrategy#getWorkerCategoryCapacity
4. In CoordinatorDutyUtils#getTotalWorkerCapacity if overlordClient.getTotalWorkerCapacity() return category based capacity, return  Math.min(totalCapacity, (int) (totalWorkerCapacity * taskSlotRatio)); here totalCapacity is category capacity.
Summary: according to the WorkerSelectStrategy(category based or not), set the category capacity in TotalWorkerCapacityResponse, in client side, compare the category capacity and totalWorkerCapacity * taskSlotRatio, take the smaller one


#### Fixed the bug 15847

#### Release note
##### Fix the bug of getCompactionTaskCapacity which is not category aware
##### Update the getTotalWorkerCapacity of OverlordResource
if WorkerSelectStrategy is category based strategy(FillCapacityWithCategorySpecWorkerSelectStrategy and EqualDistributionWithCategorySpecWorkerSelectStrategy),  TotalWorkerCapacityResponse will contains the capacity group by category.

<hr>

##### Key changed/added classes in this PR
 * `CategoryCapacityInfo`
 * `OverlordResource`
 * `TotalWorkerCapacityResponse`
 * `EqualDistributionWithCategorySpecWorkerSelectStrategy`
 * `FillCapacityWithCategorySpecWorkerSelectStrategy`
 * `WorkerSelectStrategy`
 * `WorkerSelectUtils`
 * `OverlordResourceTest`
 * `EqualDistributionWithCategorySpecWorkerSelectStrategyTest`
 * `FillCapacityWithCategorySpecWorkerSelectStrategyTest`
 * `IndexingCategoryCapacityInfo`
 * `IndexingTotalWorkerCapacityInfo`
 * `IndexingWorker`
 * `CompactSegments`
 * `CoordinatorDutyUtils`
 * `KillUnusedSegments`
 * `OverlordClientImplTest`
 * `CompactSegmentsTest`
 * `KillUnusedSegmentsTest`
<hr>

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist below are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

This PR has:

- [x] been self-reviewed.
- [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

kfaraz commented:
@AlbericByte , thanks for the changes. Please add an explanation of the approach you have taken in the PR description and also include a release note for the new API changes.

AlbericByte commented:
> @AlbericByte , thanks for the changes. Please add an explanation of the approach you have taken in the PR description and also include a release note for the new API changes.

i added the explanation and updated the release note section. for release note section, it is my first time, i am not sure it is correct format, if not, could you help to share the right format, i will update asap. 

And I will fix the test

AlbericByte commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/CategoryCapacityInfo.java:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  @JsonProperty("taskType")
+  public List<String> getTaskTypeList()
```
changed to immutableList already

AlbericByte commented:
@kfaraz @abhishekagarwal87 and @gianm 
for the failing test case, i have try sever times in my local environment using the latest code, it can successfully. do i need to increase the java memory for test, welcome for any suggestion?
```
org.apache.druid.indexing.common.task.batch.parallel.ParallelIndexSupervisorTaskResourceTest.testAPIs
Error:    Run 1: ParallelIndexSupervisorTaskResourceTest.testAPIs:214 » TestTimedOut test timed out after 20000 milliseconds
Error:    Run 2: ParallelIndexSupervisorTaskResourceTest.testAPIs:214 » TestTimedOut test timed out after 20000 milliseconds
```

AlbericByte commented:
hi @kfaraz  @abhishekagarwal87  and @gianm 
Could you help to review this pr when you are not busy, i hope i can contribute more with your help.

jakubmatyszewski commented:
unstale

jakubmatyszewski commented:
unstale

jakubmatyszewski commented:
unstale

gianm commented:
Sorry, this one fell through the cracks. Are you still interested in working on it @AlbericByte? If so, please let me know, I will do a review.
